File: sync-your-cookie/tools/cookie_encryptor.py
"""
CookieEncryptor —— Cookie 数据加密/解密工具类 (Python 版)
=========================================================

封装 sync-your-cookie 的完整编解码管线:
    CookiesMap → protobuf → raw-deflate → gzip → base64 → [AES-256-GCM → base64]

依赖:
    pip install cryptography protobuf

用法:
    from cookie_encryptor import CookieEncryptor

    # 解密（从云端拉取后还原 cookie）
    encryptor = CookieEncryptor(password="my-secret")
    cookies_map = encryptor.decode(encoded_string)

    # 加密
    encoded = encryptor.encode(cookies_map)

    # 不加密（仅压缩编码）
    no_encrypt = CookieEncryptor()
    compressed = no_encrypt.encode(cookies_map)
"""

import base64
import gzip
import json
import struct
import zlib
from typing import Any, Dict, List, Optional, Union

# ------- AES & KDF -----------------------------------------------------------
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes

# ------- Protobuf dynamic loading ---------------------------------------------
from google.protobuf import descriptor_pb2, descriptor_pool, message_factory, json_format
from google.protobuf.message import Message

# =====================================================================
# 内置 FileDescriptorSet — 等同于 cookie.proto 的编译结果，无需外部文件
# =====================================================================
_FDS_B64 = (
    "CrEECgxjb29raWUucHJvdG8irQEKBkNvb2tpZRIMCgZkb21haW4YASgJEgoKBG5hbWUYAi"
    "gJEg0KB3N0b3JlSWQYAygJEgsKBXZhbHVlGAQoCRINCgdzZXNzaW9uGAUoCBIOCghob3N0"
    "T25seRgGKAgSFAoOZXhwaXJhdGlvbkRhdGUYBygCEgoKBHBhdGgYCCgJEg4KCGh0dHBPbm"
    "x5GAkoCBIMCgZzZWN1cmUYCigIEg4KCHNhbWVTaXRlGAsoCSIqChBMb2NhbFN0b3JhZ2VJ"
    "dGVtEgkKA2tleRgBKAkSCwoFdmFsdWUYAigJIosBCgxEb21haW5Db29raWUSEAoKY3JlYX"
    "RlVGltZRgBKAMSEAoKdXBkYXRlVGltZRgCKAMSDwoJdXNlckFnZW50GAcoCRIYCgdjb29r"
    "aWVzGAUgAygLMgcuQ29va2llEiwKEWxvY2FsU3RvcmFnZUl0ZW1zGAYgAygLMhEuTG9jYW"
    "xTdG9yYWdlSXRlbSKuAQoKQ29va2llc01hcBIQCgpjcmVhdGVUaW1lGAEoAxIQCgp1cGRh"
    "dGVUaW1lGAIoAxI5Cg9kb21haW5Db29raWVNYXAYBSADKAsyIC5Db29raWVzTWFwLkRvbW"
    "FpbkNvb2tpZU1hcEVudHJ5GkEKFERvbWFpbkNvb2tpZU1hcEVudHJ5EgkKA2tleRgBKAkS"
    "GgoFdmFsdWUYAigLMg0uRG9tYWluQ29va2llOgI4AWIGcHJvdG8z"
)


class _Proto:
    """protobuf 消息类型的延迟加载容器"""

    Cookie: type = None
    LocalStorageItem: type = None
    DomainCookie: type = None
    CookiesMap: type = None

    _loaded = False

    @classmethod
    def _load(cls):
        if cls._loaded:
            return
        fds = descriptor_pb2.FileDescriptorSet()
        fds.ParseFromString(base64.b64decode(_FDS_B64))
        pool = descriptor_pool.DescriptorPool()
        for fdp in fds.file:
            pool.Add(fdp)
        cls.Cookie = message_factory.GetMessageClass(pool.FindMessageTypeByName("Cookie"))
        cls.LocalStorageItem = message_factory.GetMessageClass(pool.FindMessageTypeByName("LocalStorageItem"))
        cls.DomainCookie = message_factory.GetMessageClass(pool.FindMessageTypeByName("DomainCookie"))
        cls.CookiesMap = message_factory.GetMessageClass(pool.FindMessageTypeByName("CookiesMap"))
        cls._loaded = True


def _ensure_proto():
    _Proto._load()


# =====================================================================
# Base64
# =====================================================================

def _b64encode(data: bytes) -> str:
    return base64.b64encode(data).decode("ascii")


def _b64decode(data: str) -> bytes:
    return base64.b64decode(data.strip())


# =====================================================================
# Gzip + Raw Deflate 压缩 / 解压
# =====================================================================

def _compress(plain: bytes) -> bytes:
    """管线: zlib-deflate → gzip（与 pako.deflate + CompressionStream('gzip') 等价）

    pako.deflate() 在 v2.x 默认输出 zlib 格式（含 78 9c 头部），
    而非 raw deflate，因此 Python 端也使用 zlib 格式以保持兼容。"""
    deflated = zlib.compress(plain, level=9)
    return gzip.compress(deflated, compresslevel=9)


def _decompress(compressed: bytes) -> bytes:
    """管线: gzip 解压 → zlib-inflate（逆操作）"""
    gzip_decoded = gzip.decompress(compressed)
    return zlib.decompress(gzip_decoded)


# =====================================================================
# AES-256-GCM 加密 / 解密
# =====================================================================

_MAGIC = b"SYCE"
_SALT_LEN = 16
_IV_LEN = 12
_KDF_ITERATIONS = 100000
_VERSION = 1


def _aes_encrypt(plain: bytes, password: str) -> bytes:
    """AES-256-GCM 加密，输出格式: MAGIC(4) + VERSION(1) + SALT(16) + IV(12) + CIPHERTEXT"""
    salt = __import__("os").urandom(_SALT_LEN)
    iv = __import__("os").urandom(_IV_LEN)
    key = _derive_key(password, salt)
    aesgcm = AESGCM(key)
    ciphertext = aesgcm.encrypt(iv, plain, None)
    result = bytearray()
    result.extend(_MAGIC)
    result.append(_VERSION)
    result.extend(salt)
    result.extend(iv)
    result.extend(ciphertext)
    return bytes(result)


def _aes_decrypt(encrypted: bytes, password: str) -> bytes:
    """AES-256-GCM 解密，输入需符合 _aes_encrypt 的输出格式"""
    if encrypted[:4] != _MAGIC:
        raise ValueError("数据不是有效的加密格式（魔数 'SYCE' 不匹配）")
    offset = 4
    version = encrypted[offset]
    offset += 1
    if version != _VERSION:
        raise ValueError(f"不支持的加密版本: {version}")
    salt = encrypted[offset : offset + _SALT_LEN]
    offset += _SALT_LEN
    iv = encrypted[offset : offset + _IV_LEN]
    offset += _IV_LEN
    ciphertext = encrypted[offset:]
    key = _derive_key(password, salt)
    aesgcm = AESGCM(key)
    try:
        return aesgcm.decrypt(iv, ciphertext, None)
    except Exception:
        raise ValueError("解密失败：密码错误或数据已损坏")


def _derive_key(password: str, salt: bytes) -> bytes:
    kdf = PBKDF2HMAC(algorithm=hashes.SHA256(), length=32, salt=salt, iterations=_KDF_ITERATIONS)
    return kdf.derive(password.encode("utf-8"))


def _is_encrypted(data: bytes) -> bool:
    return data[:4] == _MAGIC


# =====================================================================
# Protobuf ↔ 字典 互转
# =====================================================================

def _msg_to_dict(msg: Message) -> dict:
    """将 protobuf Message 转为纯 Python dict（int64 Long 自动转 int）"""
    result = json_format.MessageToDict(msg, preserving_proto_field_name=True)
    _fix_int64(result)
    return result


def _fix_int64(obj):
    """将 MessageToDict 产生的 int64 字符串转回 int"""
    if isinstance(obj, dict):
        for k, v in obj.items():
            if k in ("createTime", "updateTime") and isinstance(v, str):
                obj[k] = int(v)
            else:
                _fix_int64(v)
    elif isinstance(obj, list):
        for item in obj:
            _fix_int64(item)


def _dict_to_cookies_map(data: dict) -> Message:
    """将 dict 转为 CookiesMap protobuf Message（使用 ParseDict 自动处理 map 字段）"""
    _ensure_proto()
    return json_format.ParseDict(data, _Proto.CookiesMap())


def _protobuf_encode(data: dict) -> bytes:
    """dict → CookiesMap protobuf 二进制"""
    msg = _dict_to_cookies_map(data)
    return msg.SerializeToString()


def _protobuf_decode(binary: bytes) -> dict:
    """CookiesMap protobuf 二进制 → dict"""
    _ensure_proto()
    msg = _Proto.CookiesMap()
    msg.ParseFromString(binary)
    return _msg_to_dict(msg)


# =====================================================================
# 主类
# =====================================================================

class CookieEncryptor:
    """Cookie 数据加解密工具

    参数:
        password: 加密密码，为空时不加密仅压缩
    """

    def __init__(self, password: str = ""):
        self.password = password

    # ---------- 编码 ----------

    def encode(self, cookies_map: dict, *,
               password: Optional[str] = None,
               protobuf: bool = True) -> str:
        """CookiesMap dict → 云端存储字符串

        Args:
            cookies_map: 符合 ICookiesMap 结构的 dict
            password:    加密密码，默认使用实例密码
            protobuf:    True 则 protobuf 编码，False 则返回 JSON 字符串（不加密）

        Returns:
            base64 字符串，可直接写入 Cloudflare KV / GitHub Gist
        """
        pwd = password if password is not None else self.password

        if not protobuf:
            if pwd:
                import warnings
                warnings.warn("JSON 模式不支持加密，已忽略 password")
            return json.dumps(cookies_map, ensure_ascii=False)

        proto_binary = _protobuf_encode(cookies_map)
        compressed = _compress(proto_binary)
        inner_b64 = _b64encode(compressed)

        if pwd:
            raw = _b64decode(inner_b64)
            encrypted = _aes_encrypt(raw, pwd)
            return _b64encode(encrypted)

        return inner_b64

    # ---------- 解码 ----------

    def decode(self, encoded_data: str, *,
               password: Optional[str] = None) -> dict:
        """云端存储字符串 → CookiesMap dict

        自动检测格式：JSON / protobuf / AES-GCM 加密。
        无需手动指定数据类型。

        Args:
            encoded_data: 云端存储的原始字符串
            password:     加密密码，默认使用实例密码

        Returns:
            CookiesMap dict
        """
        pwd = password if password is not None else self.password

        if encoded_data.strip().startswith("{"):
            return json.loads(encoded_data)

        raw = _b64decode(encoded_data)

        if _is_encrypted(raw):
            if not pwd:
                raise ValueError("检测到加密数据但未提供密码，请传入 password 参数。")
            raw = _aes_decrypt(raw, pwd)

        try:
            decompressed = _decompress(raw)
            return _protobuf_decode(decompressed)
        except Exception as exc:
            raise ValueError(
                f"数据解压或解码失败: {exc}。"
                f"如果数据未使用 protobuf 编码，请使用 decode_json() 方法。"
            )

    # ---------- JSON ----------

    @staticmethod
    def decode_json(json_data: str) -> dict:
        """解码 JSON 格式的 Cookie 数据（不使用 protobuf 时）"""
        return json.loads(json_data)

    # ---------- 类型检测 ----------

    @staticmethod
    def detect_type(data: str) -> str:
        """检测编码类型

        Returns:
            'json' | 'encrypted' | 'protobuf' | 'unknown'
        """
        if not data or not isinstance(data, str):
            return "unknown"
        trimmed = data.strip()
        if trimmed.startswith("{"):
            return "json"
        try:
            raw = _b64decode(trimmed)
            if _is_encrypted(raw):
                return "encrypted"
            return "protobuf"
        except Exception:
            return "unknown"


# =====================================================================
# 便捷函数
# =====================================================================

def encrypt_cookies(cookies_map: dict, password: str = "") -> str:
    """快速加密（编码）Cookie 数据"""
    return CookieEncryptor(password=password).encode(cookies_map)


def decrypt_cookies(encoded_data: str, password: str = "") -> dict:
    """快速解密（解码）Cookie 数据"""
    return CookieEncryptor(password=password).decode(encoded_data)


# =====================================================================
# 自测（python cookie_encryptor.py）
# =====================================================================
if __name__ == "__main__":
    sample = {
        "createTime": 1700000000000,
        "updateTime": 1700000001000,
        "domainCookieMap": {
            "example.com": {
                "createTime": 1690000000000,
                "updateTime": 1700000000000,
                "userAgent": "Mozilla/5.0 (Windows NT 10.0)",
                "cookies": [
                    {"domain": ".example.com", "name": "session", "value": "abc123",
                     "path": "/", "secure": True, "httpOnly": True, "sameSite": "lax"},
                    {"domain": ".example.com", "name": "token",
                     "value": "eyJhbGciOiJIUzI1NiJ9.xxx", "path": "/",
                     "secure": True, "httpOnly": True},
                ],
                "localStorageItems": [
                    {"key": "theme", "value": "dark"},
                    {"key": "lang", "value": "zh-CN"},
                ],
            },
        },
    }

    password = "test-1234"

    # 1. 加密 + 解密
    e = CookieEncryptor(password=password)
    enc = e.encode(sample)
    assert CookieEncryptor.detect_type(enc) == "encrypted", "detectType fail"
    dec = e.decode(enc)
    assert dec["domainCookieMap"]["example.com"]["cookies"][0]["value"] == "abc123"
    print("1. 加密→解密: ✓")

    # 2. 不加密
    e2 = CookieEncryptor()
    enc2 = e2.encode(sample)
    assert CookieEncryptor.detect_type(enc2) == "protobuf", f"detectType fail: {CookieEncryptor.detect_type(enc2)}"
    dec2 = e2.decode(enc2)
    assert len(dec2["domainCookieMap"]["example.com"]["cookies"]) == 2
    print("2. 不加密压缩: ✓")

    # 3. JSON 模式
    enc3 = e2.encode(sample, protobuf=False)
    assert CookieEncryptor.detect_type(enc3) == "json"
    dec3 = CookieEncryptor.decode_json(enc3)
    assert dec3["domainCookieMap"]["example.com"]["cookies"][1]["value"] == "eyJhbGciOiJIUzI1NiJ9.xxx"
    print("3. JSON 模式: ✓")

    # 4. 便捷函数
    enc4 = encrypt_cookies(sample, password)
    dec4 = decrypt_cookies(enc4, password)
    assert dec4["domainCookieMap"]["example.com"]["localStorageItems"][0]["value"] == "dark"
    print("4. 便捷函数: ✓")

    # 5. 错误密码
    try:
        e.decode(enc, password="wrong")
        assert False, "should raise"
    except ValueError as ex:
        assert "密码错误" in str(ex) or "密钥" in str(ex)
        print(f"5. 错误密码: ✓ ({ex})")

    # 6. detectType
    assert CookieEncryptor.detect_type("") == "unknown"
    assert CookieEncryptor.detect_type("{}") == "json"
    assert CookieEncryptor.detect_type(None) == "unknown"
    print("6. detectType 边界: ✓")

    # 7. 多次往返
    for i in range(3):
        enc_x = e.encode(sample)
        dec_x = e.decode(enc_x)
        assert dec_x["domainCookieMap"]["example.com"]["cookies"][0]["value"] == "abc123"
    print("7. 多次往返: ✓")

    print("\n全部测试通过!")

File: sync-your-cookie/tools/cookie-encryptor.mjs
/**
 * CookieEncryptor — Cookie 数据加密/解密工具类
 *
 * 封装 sync-your-cookie 的完整编解码管线：
 *   CookiesMap → protobuf → deflate → gzip → base64 → [AES-256-GCM → base64]
 *
 * 用法:
 *   import { CookieEncryptor } from './cookie-encryptor.mjs';
 *
 *   // 加密（从扩展获取 cookie 数据后上传到云端）
 *   const encryptor = new CookieEncryptor({ password: 'my-secret' });
 *   const encoded = await encryptor.encode(cookiesMap);
 *
 *   // 解密（从云端拉取数据后还原 cookie）
 *   const cookiesMap = await encryptor.decode(encoded);
 *
 *   // 不加密（仅压缩编码）
 *   const noEncrypt = new CookieEncryptor();
 *   const compressed = await noEncrypt.encode(cookiesMap);
 *
 * 依赖: pako, protobufjs
 */

import pako from 'pako';
import protobuf from 'protobufjs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

// ─── 常量 ─────────────────────────────────────────────────────────────────

const __dirname = dirname(fileURLToPath(import.meta.url));
const DEFAULT_PROTO_FILE = resolve(__dirname, '..', 'packages', 'protobuf', 'proto', 'cookie.proto');

/** AES-GCM 加密魔数 "SYCE" */
const MAGIC_BYTES = new Uint8Array([0x53, 0x59, 0x43, 0x45]);
const SALT_LENGTH = 16;
const IV_LENGTH = 12;
const PBKDF2_ITERATIONS = 100000;
const ENCRYPTION_VERSION = 1;

const BASE64_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

// ─── 类型定义 (JSDoc) ─────────────────────────────────────────────────────

/**
 * @typedef {Object} ICookie
 * @property {string} [domain]
 * @property {string} [name]
 * @property {string} [storeId]
 * @property {string} [value]
 * @property {boolean} [session]
 * @property {boolean} [hostOnly]
 * @property {number} [expirationDate]
 * @property {string} [path]
 * @property {boolean} [httpOnly]
 * @property {boolean} [secure]
 * @property {string} [sameSite]
 */

/**
 * @typedef {Object} ILocalStorageItem
 * @property {string} [key]
 * @property {string} [value]
 */

/**
 * @typedef {Object} IDomainCookie
 * @property {number|Long} [createTime]
 * @property {number|Long} [updateTime]
 * @property {ICookie[]} [cookies]
 * @property {ILocalStorageItem[]} [localStorageItems]
 * @property {string} [userAgent]
 */

/**
 * @typedef {Object} ICookiesMap
 * @property {number|Long} [createTime]
 * @property {number|Long} [updateTime]
 * @property {Object<string, IDomainCookie>} [domainCookieMap]
 */

// ─── Base64 编解码 ────────────────────────────────────────────────────────

function base64Encode(bytes) {
  let result = '';
  const len = bytes.byteLength;
  const remainder = len % 3;
  const mainLen = len - remainder;

  for (let i = 0; i < mainLen; i += 3) {
    const chunk = (bytes[i] << 16) | (bytes[i + 1] << 8) | bytes[i + 2];
    result += BASE64_CHARS[(chunk & 16515072) >> 18] + BASE64_CHARS[(chunk & 258048) >> 12] +
              BASE64_CHARS[(chunk & 4032) >> 6] + BASE64_CHARS[chunk & 63];
  }
  if (remainder === 1) {
    const chunk = bytes[mainLen];
    result += BASE64_CHARS[(chunk & 252) >> 2] + BASE64_CHARS[(chunk & 3) << 4] + '==';
  } else if (remainder === 2) {
    const chunk = (bytes[mainLen] << 8) | bytes[mainLen + 1];
    result += BASE64_CHARS[(chunk & 64512) >> 10] + BASE64_CHARS[(chunk & 1008) >> 4] +
              BASE64_CHARS[(chunk & 15) << 2] + '=';
  }
  return result;
}

function base64Decode(base64) {
  const binary = atob(base64.trim());
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

// ─── Gzip 压缩/解压 ───────────────────────────────────────────────────────

async function gzipCompress(data) {
  const stream = new Blob([data]).stream().pipeThrough(new CompressionStream('gzip'));
  const chunks = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return new Uint8Array(await new Blob(chunks).arrayBuffer());
}

async function gzipDecompress(data) {
  const stream = new Blob([data]).stream().pipeThrough(new DecompressionStream('gzip'));
  const chunks = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return new Uint8Array(await new Blob(chunks).arrayBuffer());
}

// ─── AES-256-GCM 加密/解密 ───────────────────────────────────────────────

async function deriveKey(password, salt) {
  const encoder = new TextEncoder();
  const baseKey = await crypto.subtle.importKey('raw', encoder.encode(password), 'PBKDF2', false, ['deriveKey']);
  return crypto.subtle.deriveKey(
    { name: 'PBKDF2', salt, iterations: PBKDF2_ITERATIONS, hash: 'SHA-256' },
    baseKey,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt'],
  );
}

async function aesEncrypt(plainBytes, password) {
  const salt = crypto.getRandomValues(new Uint8Array(SALT_LENGTH));
  const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));
  const key = await deriveKey(password, salt);
  const ciphertext = await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, plainBytes);

  const result = new Uint8Array(4 + 1 + SALT_LENGTH + IV_LENGTH + ciphertext.byteLength);
  let offset = 0;
  result.set(MAGIC_BYTES, offset); offset += 4;
  result[offset] = ENCRYPTION_VERSION; offset += 1;
  result.set(salt, offset); offset += SALT_LENGTH;
  result.set(iv, offset); offset += IV_LENGTH;
  result.set(new Uint8Array(ciphertext), offset);
  return result;
}

async function aesDecrypt(encryptedBytes, password) {
  let offset = 0;

  const magic = encryptedBytes.slice(0, 4);
  if (!magic.every((b, i) => b === MAGIC_BYTES[i])) {
    throw new Error('数据不是有效的加密格式（魔数不匹配）');
  }
  offset += 4;

  const version = encryptedBytes[offset++];
  if (version !== ENCRYPTION_VERSION) {
    throw new Error(`不支持的加密版本: ${version}`);
  }

  const salt = encryptedBytes.slice(offset, offset + SALT_LENGTH);
  offset += SALT_LENGTH;
  const iv = encryptedBytes.slice(offset, offset + IV_LENGTH);
  offset += IV_LENGTH;
  const ciphertext = encryptedBytes.slice(offset);

  const key = await deriveKey(password, salt);
  try {
    const decrypted = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, ciphertext);
    return new Uint8Array(decrypted);
  } catch {
    throw new Error('解密失败：密码错误或数据已损坏');
  }
}

// ─── Protobuf 编解码 ─────────────────────────────────────────────────────

/**
 * @returns {Promise<protobuf.Type>}
 */
async function loadProto(protoFile) {
  const root = await protobuf.load(protoFile);
  return root.lookupType('CookiesMap');
}

/**
 * @param {ICookiesMap} cookiesMap
 * @param {protobuf.Type} [CookiesMapType]
 * @returns {Uint8Array}
 */
async function protobufEncode(cookiesMap, CookiesMapType) {
  const type = CookiesMapType || await loadProto(DEFAULT_PROTO_FILE);
  const invalid = type.verify(cookiesMap);
  if (invalid) throw new Error(`Protobuf 数据验证失败: ${invalid}`);
  const message = type.create(cookiesMap);
  return type.encode(message).finish();
}

/**
 * @param {Uint8Array} buffer
 * @param {protobuf.Type} [CookiesMapType]
 * @returns {Promise<ICookiesMap>}
 */
async function protobufDecode(buffer, CookiesMapType) {
  const type = CookiesMapType || await loadProto(DEFAULT_PROTO_FILE);
  const result = type.decode(buffer);
  return JSON.parse(JSON.stringify(result, (key, value) => {
    if (typeof value === 'object' && value !== null && typeof value.toNumber === 'function') {
      return value.toNumber();
    }
    return value;
  }));
}

// ─── 主类 ─────────────────────────────────────────────────────────────────

export class CookieEncryptor {
  /**
   * @param {Object} [options]
   * @param {string} [options.password] - 加密密码。不传则跳过 AES 加密，仅进行压缩编码
   * @param {string} [options.protoFile] - .proto 文件路径，默认使用项目内置的 cookie.proto
   */
  constructor(options = {}) {
    this.password = options.password || '';
    this.protoFile = options.protoFile || DEFAULT_PROTO_FILE;
    /** @type {protobuf.Type|null} */
    this._protoType = null;
  }

  /**
   * 加载 protobuf schema（惰性加载，首次使用时自动调用）
   * @returns {Promise<protobuf.Type>}
   */
  async _loadProto() {
    if (!this._protoType) {
      this._protoType = await loadProto(this.protoFile);
    }
    return this._protoType;
  }

  /**
   * 编码 CookiesMap 为云端存储格式。
   *
   * 管线: protobuf → deflate → gzip → base64 → [AES-GCM 加密 → base64]
   *
   * @param {ICookiesMap} cookiesMap - Cookie 数据
   * @param {Object} [options]
   * @param {string} [options.password] - 覆盖实例的密码
   * @param {boolean} [options.protobuf=true] - 是否使用 protobuf 编码（false 则返回 JSON 字符串，不加密）
   * @returns {Promise<string>} 云端存储格式的字符串
   */
  async encode(cookiesMap, options = {}) {
    const password = options.password ?? this.password;
    const useProtobuf = options.protobuf !== false;

    if (!useProtobuf) {
      if (password) {
        console.warn('[CookieEncryptor] JSON 模式不支持加密，已忽略 password 参数。请使用 protobuf 模式启用加密。');
      }
      return JSON.stringify(cookiesMap);
    }

    const protoType = await this._loadProto();
    const encoded = await protobufEncode(cookiesMap, protoType);
    const deflated = pako.deflate(encoded);
    const compressed = await gzipCompress(deflated);
    const innerBase64 = base64Encode(compressed);

    if (password) {
      const rawBytes = base64Decode(innerBase64);
      const encrypted = await aesEncrypt(rawBytes, password);
      return base64Encode(encrypted);
    }

    return innerBase64;
  }

  /**
   * 从云端存储格式解码为 CookiesMap。
   *
   * 自动检测并处理：JSON、protobuf、AES-GCM 加密。
   *
   * @param {string} encodedData - 云端存储的原始数据
   * @param {Object} [options]
   * @param {string} [options.password] - 覆盖实例的密码
   * @returns {Promise<ICookiesMap>} 解密后的 CookiesMap
   */
  async decode(encodedData, options = {}) {
    const password = options.password ?? this.password;
    let processed = encodedData;

    // JSON 格式（以 { 开头）
    if (processed.startsWith('{')) {
      return JSON.parse(processed);
    }

    // Base64 解码
    let rawBytes;
    try {
      rawBytes = base64Decode(processed);
    } catch {
      throw new Error('无法解码 base64，数据格式不正确');
    }

    // 检测并处理加密
    if (isEncrypted(rawBytes)) {
      if (!password) {
        throw new Error('检测到加密数据但未提供密码。请传入 password 参数。');
      }
      const decrypted = await aesDecrypt(rawBytes, password);
      processed = base64Encode(decrypted);
    }

    // Protobuf 解码: base64 → gzip 解压 → raw inflate → protobuf
    try {
      const compressedBytes = base64Decode(processed);
      const decompressed = await gzipDecompress(compressedBytes);
      const inflated = pako.inflate(decompressed);
      const protoType = await this._loadProto();
      return await protobufDecode(inflated, protoType);
    } catch (err) {
      throw new Error(`数据解压或解码失败: ${err.message}。如果数据未使用 protobuf 编码，请使用 decodeJson() 方法。`);
    }
  }

  /**
   * 解码 JSON 格式的 Cookie 数据（不使用 protobuf 编码的数据）。
   *
   * @param {string} jsonData - JSON 字符串
   * @returns {ICookiesMap}
   */
  decodeJson(jsonData) {
    return JSON.parse(jsonData);
  }

  /**
   * 检测数据编码类型。
   *
   * @param {string} data - 云端存储的原始数据字符串
   * @returns {'json' | 'encrypted' | 'protobuf' | 'unknown'}
   */
  static detectType(data) {
    if (!data || typeof data !== 'string') return 'unknown';

    const trimmed = data.trim();
    if (trimmed.startsWith('{')) return 'json';

    try {
      const bytes = base64Decode(trimmed);
      if (isEncrypted(bytes)) return 'encrypted';
      return 'protobuf';
    } catch {
      return 'unknown';
    }
  }

  /**
   * 重新加载 protobuf schema（切换 proto 文件后调用）。
   */
  async reloadProto() {
    this._protoType = null;
    return this._loadProto();
  }
}

/**
 * 检查 Uint8Array 是否为 AES-GCM 加密数据（检测 SYCE 魔数）。
 * @param {Uint8Array} data
 * @returns {boolean}
 */
function isEncrypted(data) {
  if (data.length < MAGIC_BYTES.length) return false;
  return data.slice(0, MAGIC_BYTES.length).every((b, i) => b === MAGIC_BYTES[i]);
}

// ─── 便捷函数 ─────────────────────────────────────────────────────────────

/**
 * 快速加密 Cookie 数据。
 * @param {ICookiesMap} cookiesMap
 * @param {string} [password] - 不传则不加密
 * @returns {Promise<string>}
 */
export async function encryptCookies(cookiesMap, password) {
  const e = new CookieEncryptor({ password });
  return e.encode(cookiesMap);
}

/**
 * 快速解密 Cookie 数据。
 * @param {string} encodedData
 * @param {string} [password]
 * @returns {Promise<ICookiesMap>}
 */
export async function decryptCookies(encodedData, password) {
  const e = new CookieEncryptor({ password });
  return e.decode(encodedData);
}
